[docs][Ray Clusters] Key Concepts page (#27510)
ray-project/ray: doc/source/cluster/cluster_under_construction/key-concepts.rst

.. contents::
    :local:

Ray cluster
------------
A **Ray cluster** is comprised of a :ref:`head node<cluster-head-node-under-construction>`
and any number of :ref:`worker nodes<cluster-worker-nodes-under-construction>`.

.. figure:: images/ray-cluster.svg
    :align: center
    :width: 600px
    
    *A Ray cluster with two worker nodes. Each node runs Ray helper processes to
    facilitate distributed scheduling and memory management. The head node runs
    additional control processes, which are highlighted.*

The number of worker nodes in a cluster may change with application demand, according
to your Ray cluster configuration. This is known as *autoscaling*. The head node runs
the :ref:`autoscaler<cluster-autoscaler-under-construction>`.

.. note::
    Ray nodes are implemented as pods when :ref:`running on Kubernetes<kuberay-index>`.

Users can submit jobs for execution on the Ray cluster, or can interactively use the
cluster by connecting to the head node and running `ray.init`. See
:ref:`Clients and Jobs<cluster-clients-and-jobs-under-construction>` for more information.

.. _cluster-worker-nodes-under-construction:

Worker nodes
~~~~~~~~~~~~

> **Worker nodes** are nodes which run Ray and are connected to the head node. Ray may schedule
> tasks and actors onto worker nodes. Each worker node runs helper processes which

Reviewer: ```suggestion
**Worker nodes** execute a Ray application by executing tasks and actors and storing Ray objects. Each worker node runs helper processes which
```
(I think this is a bit more straightforward).

implement distributed scheduling and :ref:`memory management<memory>`.

.. _cluster-head-node-under-construction:

Head node
~~~~~~~~~
Every Ray cluster has one node which is designated as the **head node** of the cluster.

> The head node runs singleton processes responsible for cluster management such as the

Reviewer: ```suggestion
The head node is identical to other worker nodes, except that it also runs singleton processes responsible for cluster management such as the
```

:ref:`autoscaler<cluster-autoscaler-under-construction>` and the Ray driver processes
:ref:`which run Ray jobs<cluster-clients-and-jobs-under-construction>`. Ray may schedule
tasks and actors on the head node just like any other worker node, unless configured otherwise.

> .. note::

Reviewer: Do we need this note, @DmitriGekhtman? It seems pretty off-topic / too much detail for a first-time user.

Reviewer: Alternatively, we could move it as an inline note to the "Autoscaler" subsection (suggestion below).

Author: I committed your suggestion and removed this note!

    :ref:`On Kubernetes<kuberay-index>`, the autoscaler process runs inside a sidecar container in the head pod,
    instead of in the same container (which would be most analogous to how the autoscaler runs inside the head node
    on non-Kubernetes clusters).

.. _cluster-autoscaler-under-construction:

Autoscaler
----------

> The **autoscaler** is a process that runs on the :ref:`head node<cluster-head-node-under-construction>`.

Reviewer: ```suggestion
The **autoscaler** is a process that runs on the :ref:`head node<cluster-head-node-under-construction>` (or as a sidecar container in the head pod if :ref:`using Kubernetes<kuberay-index>`).
```

.. include:: /_includes/clusters/we_are_hiring.rst

Key Concepts
============

.. _cluster-key-concepts-under-construction:

This page introduces the following key concepts, an understanding of which is helpful when creating and managing Ray clusters:

.. contents::
    :local:

.. _cluster-worker-nodes-under-construction:

Worker nodes
------------
A Ray cluster consists of a set of one or more **worker nodes**. Ray tasks and actors
can be scheduled onto the worker nodes. Each worker node runs helper processes which
implement distributed scheduling and :ref:`memory management<memory>`.

.. figure:: ray-cluster.jpg
    :align: center
    :width: 600px

>     A Ray cluster consists of a set of one or more worker nodes, one of which is designated

Reviewer 2: I'd remove the word "worker" in this sentence.
We usually categorize workers and head into disjoint buckets.

Author: I've intentionally removed the notion that they're disjoint buckets. The distinction necessitates explanation on why they're different (e.g. `Node Types`), whereas here a head node is simply a worker node plus additional things.

Thoughts?

This isn't original btw, I was inspired by the [simple explanations in the whitepaper](https://docs.google.com/document/d/1lAy0Owi-vPz2jEqBSaHNQcy2IBSDEHyXNOQZlGuj93c/preview#heading=h.y4i8cfntvi1k) :)

Reviewer 2: Oh I guess that is how the whitepaper explains it.

Well, they're different in that a head node has additional things.
Often the head node requires different configuration than worker nodes. (This is the case in our sample configs.)
For large workloads we often recommend hacking the head node with num-cpus=0 to avoid scheduling Ray workloads there and make it more-or-less into a control node.

If the head node is a special kind of worker node, I guess "worker" is redundant -- we just have nodes, one of which is a head. Eliminating the terminology "worker node" altogether would also reduce the conflation between "worker nodes" and "worker processes", but I think this ship has sailed.

Anyways, I think I'll defer to the authorities on this one :)

Reviewer 2: In an ideal state, there would be no head node -- just worker nodes, and some control processes which can run wherever.

Reviewer: +1 to @DmitriGekhtman, let's remove the term "worker node".

Reviewer 3: My 2c: 
- it seems that it's better to keep the "cluster" as a key concept. We could emphasize that ray cluster just include one or more nodes and explain briefly the distinction between head node and worker node. After this brief explanation, we could use "head node" and "worker node" in the rest of the doc.
- we should revamp the diagram to highlight the differences between the head node and worker node precisely and concisely.
- The note "Here, a node means an individual machine. However, [on Kubernetes](https://ray--27510.org.readthedocs.build/en/27510/cluster/cluster_under_construction/ray-clusters-on-kubernetes/index.html#kuberay-index), nodes are implemented as pods." could also be moved under Ray cluster.

------- here is how it might look like ------

### Ray cluster
A Ray cluster consists of one or more nodes, one of which is designated the head node while the rest are often called worker nodes.
<img width="833" alt="Screen Shot 2022-08-05 at 1 33 48 PM" src="https://user-images.githubusercontent.com/9677264/183158561-d29a82aa-56cd-42c2-8a03-f687f86d30a2.png">

The number of worker nodes in a cluster may change with application demand, according to your Ray cluster configuration. The [head node](https://ray--27510.org.readthedocs.build/en/27510/cluster/cluster_under_construction/key-concepts.html#cluster-head-node-under-construction) runs the logic which implements autoscaling.

<img width="775" alt="Screen Shot 2022-08-05 at 1 39 51 PM" src="https://user-images.githubusercontent.com/9677264/183159300-897f070d-4f6e-4cdd-8230-05639a0a281c.png">

Author: Thanks for the feedback, everyone! This is implemented.

    as the head node. Each node runs Ray helper processes and Ray application code.

The number of worker nodes in a cluster may change with application demand, according
to your Ray cluster configuration. The :ref:`head node<cluster-head-node-under-construction>`
runs the logic which implements autoscaling.

.. note::

>     Here, a *node* means an individual machine. However, :ref:`on Kubernetes<kuberay-index>`, nodes are implemented as pods.

Reviewer 2: I'd say `Ray nodes are implemented as pods.`

.. _cluster-head-node-under-construction:

Head node
---------

> Every Ray cluster has one :ref:`worker node<cluster-worker-nodes-under-construction>`

Reviewer 2: remove "worker" for same reason as above

Author: Done

which is designated as the **head node** of the cluster. The head node runs
important processes such as the :ref:`autoscaler<cluster-autoscaler-under-construction>`

> and the Ray driver processes, :ref:`which run the top-level Ray application

Reviewer 2: I think these days job drivers may actually land on worker nodes.
@brucez-anyscale or @simon-mo to comment

Reviewer 4: Not sure about job drivers. What I do is move `dashboard agent` which has a serve api http server to raylet.

Reviewer 5: no yet. job drivers are still tied to head node at the moment. this is part of @architkulkarni's future jobs GA plan

<cluster-clients-and-jobs-under-construction>`. Ray may schedule tasks and actors
on the head node just like any other worker node.

.. _cluster-autoscaler-under-construction:

Autoscaler
----------

The autoscaler is a process that runs on the :ref:`head node<cluster-head-node-under-construction>`.
It is responsible for provisioning or deprovisioning :ref:`worker nodes<cluster-worker-nodes-under-construction>`
to meet the needs of the Ray workload. In particular, if the resource demands of the Ray workload exceed the
current capacity of the cluster, the autoscaler will attempt to add more nodes. Conversely, if
a node is idle for long enough, the autoscaler will remove it from the cluster.

To learn more about the autoscaler and how to configure it, refer to the following user guides:

* :ref:`Configuring Autoscaling on VMs<deployment-guide-autoscaler-under-construction>`.

> * :ref:`Configuring Autoscaling on Kubernetes<kuberay-index>` (TODO cade@ update this link.)

Reviewer 2: It's a section in this PR :)
https://github.com/ray-project/ray/pull/27504

.. _cluster-clients-and-jobs-under-construction:

Clients and Jobs
----------------
Ray provides two methods for running workloads on a Ray Cluster: the Ray Client, and Ray Job Submission.

* **The Ray Client** enables interactive development by connecting a local Python script or shell to the cluster.
  Developers can scale-out their local programs on the cloud as if it were on their laptop. The Ray Client is used
  by specifying the :ref:`head node<cluster-head-node-under-construction>` address as an argument to `ray.init`.
* **Ray Job Submission** enables users to submit locally developed-and-tested applications to a remote Ray
  Cluster. Ray Job Submission simplifies the experience of packaging, deploying, and managing a Ray application.

To learn how to run workloads on a Ray Cluster, refer to the following user guides:

* :ref:`Running Ray workloads on VMs<ref-cluster-quick-start-vms-under-construction>` (TODO cade@ update this link).

> * :ref:`Running Ray workloads on Kubernetes<kuberay-index>` (TODO cade@ update this link).

Reviewer 2: There's a quickstart guide that can be linked `<kuberay-quickstart>`.
Linking the K8s landing page <kuberay-index> is ok to.

.. include:: /_includes/clusters/we_are_hiring.rst

Key Concepts
============

.. _cluster-key-concepts-under-construction:

> This page introduces the following key concepts, an understanding of which is helpful when creating and managing Ray clusters:

Reviewer: ```suggestion
This page introduces the following key concepts concerning Ray clusters:
```

.. contents::
    :local:

.. _cluster-worker-nodes-under-construction:

Worker nodes
------------

> A Ray cluster consists of a set of one or more **worker nodes**. Ray tasks and actors

Author: Done

> can be scheduled onto the worker nodes. Each worker node runs helper processes which

Reviewer 2: I think this is up-to-taste in an overview, since a placement group is a mechanism for task/actor scheduling.

> can be scheduled onto the worker nodes. Each worker node runs helper processes which

Author: I am of same opinion as @DmitriGekhtman , lmk if you want to discuss more @clarng !

implement distributed scheduling and :ref:`memory management<memory>`.

.. figure:: ray-cluster.jpg
    :align: center
    :width: 600px

    A Ray cluster consists of a set of one or more worker nodes, one of which is designated
    as the head node. Each node runs Ray helper processes and Ray application code.

The number of worker nodes in a cluster may change with application demand, according
to your Ray cluster configuration. The :ref:`head node<cluster-head-node-under-construction>`
runs the logic which implements autoscaling.

.. note::
    Here, a *node* means an individual machine. However, :ref:`on Kubernetes<kuberay-index>`, nodes are implemented as pods.

.. _cluster-head-node-under-construction:

Head node
---------
Every Ray cluster has one :ref:`worker node<cluster-worker-nodes-under-construction>`
which is designated as the **head node** of the cluster. The head node runs

> important processes such as the :ref:`autoscaler<cluster-autoscaler-under-construction>`

Reviewer: ```suggestion
singleton processes responsible for cluster management such as the :ref:`autoscaler<cluster-autoscaler-under-construction>`
```

and the Ray driver processes, :ref:`which run the top-level Ray application
<cluster-clients-and-jobs-under-construction>`. Ray may schedule tasks and actors
on the head node just like any other worker node.

.. note::
    :ref:`On Kubernetes<kuberay-index>`, the autoscaler process runs inside a sidecar container in the head pod,
    instead of in the same container (which would be most analogous to how the autoscaler runs inside the head node

>     on non-Kubernetes clusters).

Reviewer 2: I think this wording is ok.
On VMs, we typically also recommend running in a Docker container, in which it case it is literally the case that the autoscaler runs in the same container as the rest of the Ray process.
(It is also possible to run on the bare VM with no Docker container.)

.. _cluster-autoscaler-under-construction:

Autoscaler
----------

The autoscaler is a process that runs on the :ref:`head node<cluster-head-node-under-construction>`.
It is responsible for provisioning or deprovisioning :ref:`worker nodes<cluster-worker-nodes-under-construction>`
to meet the needs of the Ray workload. In particular, if the resource demands of the Ray workload exceed the
current capacity of the cluster, the autoscaler will attempt to add more nodes. Conversely, if
a node is idle for long enough, the autoscaler will remove it from the cluster.

To learn more about the autoscaler and how to configure it, refer to the following user guides:

* :ref:`Configuring Autoscaling on VMs<deployment-guide-autoscaler-under-construction>`.
* :ref:`Configuring Autoscaling on Kubernetes<kuberay-index>` (TODO cade@ update this link.)

.. _cluster-clients-and-jobs-under-construction:

> Clients and Jobs

Reviewer 3: This section is not very accurate after we consolidate the concept of ray jobs (see [design doc](https://docs.google.com/document/d/1Odz5cpev-dGC96fkmos5v08AIs2PTk6sJh_mfmRx1Gs/edit#) here).

Ray jobs include driver jobs and submission jobs.
- Driver jobs are created via ray client or running ray scripts locally
- Submission jobs are created via submission APIs

-------- Here is how it might look like -------
**Ray jobs**

Ray supports two types of jobs for running workloads on a Ray cluster: Driver jobs and Submission jobs.

* **Driver jobs** are created via ray client, which enables interactive development by connecting a local Python script or shell to the cluster....

* **Submission jobs** are created via submission APIs, which enables users to submit locally developed-and-tested applications to a remote Ray...

@alanwguo has more insights into this. Could you review this part?

Author: I sync'd with @alanwguo , will update names!

----------------
Ray provides two methods for running workloads on a Ray Cluster: the Ray Client, and Ray Job Submission.

> * **The Ray Client** enables interactive development by connecting a local Python script or shell to the cluster.

Reviewer: A few things:
1. We should explain that it's also fine to just SSH to the cluster and run `ray.init`. I think this should go first since I imagine a lot of people will want to do this when they're trying things out.
2. We should probably de-emphasize Ray Client in favor of Ray jobs.

Basically, we should think about what the user's flow is going from laptop to cluster and try to order the docs based on that. For example:
developing local script on laptop -> set up a cluster -> try out small version of script on cluster -> deploy full script on cluster

Reviewer 2: can mention "ssh into head node or (kubectl exec into ray head pod on kubernetes)"

.. include:: /_includes/clusters/we_are_hiring.rst

Key Concepts
============

.. _cluster-key-concepts-under-construction:

This page introduces the following key concepts concerning Ray clusters:

.. contents::
    :local:

Ray cluster
------------
A **Ray cluster** is comprised of a :ref:`head node<cluster-head-node-under-construction>`
and any number of :ref:`worker nodes<cluster-worker-nodes-under-construction>`.

.. figure:: images/ray-cluster.svg
    :align: center
    :width: 600px
    
    *A Ray cluster with two worker nodes. Each node runs Ray helper processes to
    facilitate distributed scheduling and memory management. The head node runs

>     additional helper processes, which are highlighted.*

Reviewer 2: Maybe "additional control processes"?

The number of worker nodes in a cluster may change with application demand, according
to your Ray cluster configuration. This is known as *autoscaling*. The head node runs
the :ref:`autoscaler<cluster-autoscaler-under-construction>`.

.. note::

>     Nodes are implemented as pods when :ref:`running on Kubernetes<kuberay-index>`.

Reviewer 2: Since Kubernetes has its own node concept, I'd say `Ray nodes are implemented as pods...`

Users can submit jobs for execution on the Ray cluster, or can interactively use the

> cluster by connecting to the head node and running `ray.init`. See

Reviewer 2: Could be even more specific here and include the exact function call: `ray.init()`.

Author: This feels like a lot of detail for the introduction of the Key Concepts page. How about I link to the below section `Clients and jobs` and provide this detail? FYI still incorporating Stephanie's feedback for that section.

Reviewer 2: The amount of added detail I'm asking for is the parens.

Reviewer 2: Oh, I guess `connecting to the head node` is a little vague (might be referring generally to Ray Client or SSHing)
Adding detail elsewhere is fine. Don't mind me.

:ref:`Clients and Jobs<cluster-clients-and-jobs-under-construction>` for more information.

.. _cluster-worker-nodes-under-construction:

Worker nodes
~~~~~~~~~~~~
**Worker nodes** are nodes which run Ray and are connected to the head node. Ray may schedule
tasks and actors onto worker nodes. Each worker node runs helper processes which
implement distributed scheduling and :ref:`memory management<memory>`.

.. _cluster-head-node-under-construction:

Head node
~~~~~~~~~
Every Ray cluster has one node which is designated as the **head node** of the cluster.
The head node runs singleton processes responsible for cluster management such as the
:ref:`autoscaler<cluster-autoscaler-under-construction>` and the Ray driver processes
:ref:`which run Ray jobs<cluster-clients-and-jobs-under-construction>`. Ray may schedule

> tasks and actors on the head node just like any other worker node unless configured otherwise.

Reviewer 2: nit: comma
`worker node, unless configured otherwise.`
